File: python/Trees/serialize_deserialize_binary_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Codec:
    """
    Codec for serializing and deserializing a binary tree using preorder traversal.
    serialize: O(n) time, O(n) space
    deserialize: O(n) time, O(n) space
    """
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        res = []
        def dfs(node):
            if not node:
                res.append("N")
                return
            res.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
        dfs(root)
        return ",".join(res)
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        values = data.split(",")
        self.i = 0
        def dfs():
            if values[self.i] == "N":
                self.i += 1
                return None
            node = TreeNode(int(values[self.i]))
            self.i += 1
            node.left = dfs()
            node.right = dfs()
            return node
        return dfs()

# -------------------
# Helper Functions for Testing
# -------------------
def build_tree_from_list(lst):
    if not lst:
        return None
    nodes = [TreeNode(val) if val is not None else None for val in lst]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids: node.left = kids.pop()
            if kids: node.right = kids.pop()
    return root

def tree_to_list(root):
    if not root:
        return []
    result = []
    queue = [root]
    while any(queue):
        node = queue.pop(0)
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result

def is_same_tree(p, q):
    if not p and not q:
        return True
    if not p or not q or p.val != q.val:
        return False
    return is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)

# -------------------
# Test Cases
# -------------------
def test_serialize_deserialize():
    codec = Codec()
    # Test case 1: Example
    root = build_tree_from_list([1,2,3,None,None,4,5])
    data = codec.serialize(root)
    new_root = codec.deserialize(data)
    assert is_same_tree(root, new_root)
    # Test case 2: Single node
    root = build_tree_from_list([1])
    data = codec.serialize(root)
    new_root = codec.deserialize(data)
    assert is_same_tree(root, new_root)
    # Test case 3: Empty tree
    root = build_tree_from_list([])
    data = codec.serialize(root)
    new_root = codec.deserialize(data)
    assert is_same_tree(root, new_root)
    # Test case 4: Left-skewed tree
    root = build_tree_from_list([1,2,None,3])
    data = codec.serialize(root)
    new_root = codec.deserialize(data)
    assert is_same_tree(root, new_root)
    # Test case 5: Right-skewed tree
    root = build_tree_from_list([1,None,2,None,3])
    data = codec.serialize(root)
    new_root = codec.deserialize(data)
    assert is_same_tree(root, new_root)
    print("All test cases passed!")

if __name__ == "__main__":
    test_serialize_deserialize()

File: C++/Trees/serialize_deserialize_binary_tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Codec {
public:
    /**
     * Serializes a tree to a single string.
     * 
     * Approach:
     * - Use preorder DFS traversal.
     * - If a node is null, append "N" to the string.
     * - Otherwise, append its value and recursively process left and right children.
     * 
     * Time: O(n) — visits every node once.
     * Space: O(n) — stores the result string and recursion stack.
     */
    string serialize(TreeNode* root) {
        string res;
        dfsSerialize(root, res);
        if (!res.empty() && res.back() == ',') res.pop_back(); // remove trailing comma
        return res;
    }

    /**
     * Deserializes a string to a tree.
     * 
     * Approach:
     * - Split the string on commas into tokens.
     * - Use a helper function to construct the tree recursively.
     * - Increment index as you consume tokens.
     * 
     * Time: O(n) — visits every node once.
     * Space: O(n) — recursion stack + tokens vector.
     */
    TreeNode* deserialize(string data) {
        vector<string> tokens;
        stringstream ss(data);
        string token;

        while (getline(ss, token, ',')) {
            tokens.push_back(token);
        }

        int index = 0;
        return dfsDeserialize(tokens, index);
    }

private:
    void dfsSerialize(TreeNode* node, string &res) {
        if (!node) {
            res += "N,";
            return;
        }
        res += to_string(node->val) + ",";
        dfsSerialize(node->left, res);
        dfsSerialize(node->right, res);
    }

    TreeNode* dfsDeserialize(vector<string> &tokens, int &i) {
        if (i >= tokens.size() || tokens[i] == "N") {
            i++;
            return nullptr;
        }
        TreeNode* node = new TreeNode(stoi(tokens[i++]));
        node->left = dfsDeserialize(tokens, i);
        node->right = dfsDeserialize(tokens, i);
        return node;
    }
};

/*
-------------------- LINE-BY-LINE SYNTAX EXPLANATION --------------------

#include <bits/stdc++.h>
    ✅ Includes all standard C++ headers.

using namespace std;
    ✅ Avoids prefixing standard library names with std::.

struct TreeNode { ... }
    ✅ Defines a binary tree node with int val, and left & right pointers.

TreeNode() : val(0), left(nullptr), right(nullptr) {}
    ✅ Default constructor initializes val=0 and children to nullptr.

TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    ✅ Constructor that sets val=x.

TreeNode(int x, TreeNode *left, TreeNode *right)
    ✅ Constructor that sets val=x, left, and right children.

class Codec { ... }
    ✅ Defines the Codec class with public & private methods.

string serialize(TreeNode* root)
    ✅ Public method that serializes tree into a string.

dfsSerialize(root, res);
    ✅ Helper method that appends node values to res string.

if (!node) { res += "N,"; return; }
    ✅ If node is null, add "N" and stop.

res += to_string(node->val) + ",";
    ✅ Append node’s value as string.

TreeNode* deserialize(string data)
    ✅ Public method to deserialize a string back into tree.

stringstream ss(data);
    ✅ Create stream to split string by ','.

getline(ss, token, ',')
    ✅ Read tokens one by one until end.

int index = 0;
    ✅ Start at first token.

dfsDeserialize(tokens, index);
    ✅ Build tree recursively.

if (tokens[i] == "N")
    ✅ If token is "N", return nullptr.

TreeNode* node = new TreeNode(stoi(tokens[i++]));
    ✅ Create new node from integer value.

node->left = dfsDeserialize(tokens, i);
node->right = dfsDeserialize(tokens, i);
    ✅ Build left & right children recursively.

return node;
    ✅ Return the rebuilt tree node.

*/
